### PR TITLE
Migrates to DeviceKey and new models

### PR DESCRIPTION
**Pull request by ata-peppered**
- SecureEnclave helpers
- DeviceKey stored in the enclave
- DeviceKey encryption, decryption and signing
- ApprovalTypes V2
- async signature producing
- Multiple signature types
- Photo capture and device registration

Note that the UI and the details of each approval type is incomplete

### CODE REVIEW

File: Censo/Models/ApprovalRequest.swift [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    let submitDate: Date
+    let approvalTimeoutInSeconds: Int?
+    let numberOfDispositionsRequired: Int
+    let numberOfApprovalsReceived: Int
+    let numberOfDeniesReceived: Int
+    let vaultName: String?
+    let initiationOnly: Bool
+    let details: ApprovalRequestType
+}
+
```
Comment (ata-peppered) [inline Censo/Models/ApprovalRequest.swift:23]: All request types are defined here

File: Censo/Models/ApprovalRequestType+Language.swift [hunk 2]
```diff
@@ -0,0 +1,91 @@
+//
+//  ApprovalRequestType+Language.swift
+//  Censo
+//
+//  Created by Ata Namvari on 2022-05-18.
+//
+
+import Foundation
+import SwiftUI
+
```
Comment (ata-peppered) [inline Censo/Models/ApprovalRequestType+Language.swift:10]: Headers and subheaders as they appear in the list are defined here for each type

File: Censo/Models/ClientModels/BitcoinClientModel.swift [hunk 3]
```diff
@@ -0,0 +1,36 @@
+//
+//  BitcoinClientModel.swift
+//  Censo
+//
+//  Created by Ata Namvari on 2023-01-30.
+//
+
+import Foundation
+
```
Comment (ata-peppered) [inline Censo/Models/ClientModels/BitcoinClientModel.swift:9]: Bitcoin specific models

File: Censo/Models/ClientModels/EtheriumClientModel.swift [hunk 4]
```diff
@@ -0,0 +1,148 @@
+//
+//  EtheriumClientModel.swift
+//  Censo
+//
+//  Created by Ata Namvari on 2023-01-30.
+//
+
+import Foundation
+
```
Comment (ata-peppered) [inline Censo/Models/ClientModels/EtheriumClientModel.swift:9]: Ethereum specific models

File: Censo/Models/ClientModels/SharedClientModel.swift [hunk 5]
```diff
@@ -0,0 +1,357 @@
+//
+//  SolanaClientModel.swift
+//  Censo
+//
+//  Created by Ata Namvari on 2022-03-07.
+//
+
+import Foundation
+
```
Comment (ata-peppered) [inline Censo/Models/ClientModels/SharedClientModel.swift:9]: Shared models across different chains

File: Censo/Models/Signing/ApprovalDispositionRequest+Signing.swift [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    func privateKeys() throws -> PrivateKeys {
+        let rootSeed = try deviceKey.decrypt(data: encryptedRootSeed)
+        return try PrivateKeys(rootSeed: rootSeed.bytes)
+    }
+
+    func devicePublicKey() throws -> String {
+        try deviceKey.publicExternalRepresentation().base58String
+    }
+}
+
```
Comment (ata-peppered) [inline Censo/Models/Signing/ApprovalDispositionRequest+Signing.swift:33]: Signing happens here for each approval type asynchronously

File: Censo/Models/Signing/EtheriumSignable.swift [hunk 7]
```diff
@@ -0,0 +1,242 @@
+//
+//  EtheriumSignable.swift
+//  Censo
+//
+//  Created by Ata Namvari on 2023-01-19.
+//
+
+import Foundation
+
```
Comment (ata-peppered) [inline Censo/Models/Signing/EtheriumSignable.swift:9]: Ethereum Signing

File: Censo/Models/Signing/SignatureInfo.swift [hunk 8]
```diff
@@ -0,0 +1,46 @@
+//
+//  SignatureInfo.swift
+//  Censo
+//
+//  Created by Ata Namvari on 2023-01-11.
+//
+
+import Foundation
+
```
Comment (ata-peppered) [inline Censo/Models/Signing/SignatureInfo.swift:9]: Different signature types

File: Censo/KeyStorage/Keychain+Storage.swift [hunk 9]
```diff
@@ -16,49 +16,53 @@ extension JSONEncoder {
 }
 
 extension Keychain {
-    static private let rootSeedService = "com.censocustody.root-seed-biometry"
+    static private let rootSeedService = "com.censocustody.root-seed-encrypted"
```
Comment (bflood-strike) [inline Censo/KeyStorage/Keychain+Storage.swift:19]: so does this mean that folks who had a root seed saved before (like for prerelease) will now have a new one generated or is the old root seed migrated to the new location

File: Censo/Views/Registration/KeyConfirmationSuccess.swift [hunk 10]
```diff
@@ -90,38 +95,41 @@ struct KeyConfirmationSuccess: View {
...
+                chain: .censo
             )
         ]
+
+        let bytes = signers
+            .map(\.publicKey)
+            .map { key in
+                Base58.decode(key)
+            }
+            .flatMap {
```
Comment (bflood-strike) [inline Censo/Views/Registration/KeyConfirmationSuccess.swift:120]: nice

**Review by bflood-strike — APPROVED** [review]
looks good to me - a few minor comments - nice job.

File: Censo/Models/ClientModels/SharedClientModel.swift [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+struct AccountInfo: Codable, Equatable {
+    let name: String
+    let identifier: String
+    let accountType: AccountType
+    let address: String
+    let chain: Chain
+}
+
+struct TransactionInput: Codable, Equatable {
```
[the remark this answers is not in the document]
Comment (ata-peppered) [reply Censo/Models/ClientModels/SharedClientModel.swift:101]: moved

File: Censo/Models/ClientModels/SharedClientModel.swift [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+}
+
+struct SignerInfo: Codable, Equatable {
+    let publicKey: String
+    let name: String
+    let email: String
+    let nameHashIsEmpty: Bool
+}
+
+struct SolanaDApp: Codable, Equatable {
```
[the remark this answers is not in the document]
Comment (ata-peppered) [reply Censo/Models/ClientModels/SharedClientModel.swift:124]: removed

File: Censo/Views/ApprovalRequests/ApprovalRequestItem.swift [hunk 13]
```diff
@@ -12,109 +12,122 @@ import Alamofire
...
     }
 }
+
+protocol WithdrawalRequest {
+    var account: AccountInfo { get }
+    var symbolAndAmountInfo: SymbolAndAmountInfo { get }
+    var destination: DestinationAddress { get }
+}
+
+extension BitcoinWithdrawalRequest: WithdrawalRequest {}
```
[the remark this answers is not in the document]
Comment (ata-peppered) [reply Censo/Views/ApprovalRequests/ApprovalRequestItem.swift:117]: Fixed

File: Censo/Views/Registration/KeyConfirmationSuccess.swift (hunk 10 above)
Comment (ata-peppered) [reply Censo/Views/Registration/KeyConfirmationSuccess.swift:120]: Swift is nice :)

File: Censo/Models/ClientModels/SharedClientModel.swift [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+            numberOfApprovalsReceived: 1,
+            numberOfDeniesReceived: 1,
+            vaultName: "Test Vault",
+            initiationOnly: false,
+            details: .bitcoinWithdrawalRequest(.sample)
+        )
+    }
+}
+
+extension BitcoinWithdrawalRequest {
```
[the remark this answers is not in the document]
Comment (ata-peppered) [reply Censo/Models/ClientModels/SharedClientModel.swift:215]: moved

**Pull request merged**